File: python/_async_extract_doc.py
#!/usr/local/bin/python

'''
This script performs an asynchronous extraction of the supplied PDF. See
https://docs.sensible.so/docs/api-tutorial-async-1 for more details
'''

import requests
import time
import json

# The name of a document type in Sensible, e.g., auto_insurance_quote
DOCUMENT_TYPE = "YOUR_DOCUMENT_TYPE"
# The URL of the PDF you'd like to parse
DOCUMENT_URL = "YOUR_PDF_URL"
# Your Sensible API key
API_KEY = "YOUR_API_KEY"

def extract_from_doc_url():  
  body = json.dumps({"document_url": DOCUMENT_URL})
  headers = {
    'Authorization': 'Bearer {}'.format(API_KEY),  'Content-Type': 'application/json'
  }
  response  = requests.request("POST", "https://api.sensible.so/v0/extract_from_url/{}".format(DOCUMENT_TYPE),
    headers=headers,
    data=body)
  try:
    response.raise_for_status()
  except requests.RequestException as err:
    print(response.text)
  # This is the ID we'll poll to retrieve the extraction
  # In production you'd use a webhook to avoid polling
  id = response.json()['id']
  json_response = {}
  poll_count = 0
  while not "parsed_document" in json_response:
    # Wait a few seconds for the extraction to complete on each iteration
    time.sleep(3)
    poll_count +=1
    headers = {
      'Authorization': 'Bearer {}'.format(API_KEY)
    }
    response  = requests.request("GET", "https://api.sensible.so/v0/documents/{}".format(id),
      headers=headers)
    try:
      response.raise_for_status()
    except requests.RequestException as err:
      print(response.text)
      break
    json_response = response.json()
    print("Poll attempt: {} status: {}".format(poll_count, json_response["status"]))
    if json_response["status"] == "FAILED":
      break
  print(json.dumps(json_response, indent=2))


if __name__ == '__main__':
  extract_from_doc_url()

File: node/_async-extract-doc.js
#!/usr/bin/env node

var fetch = require('isomorphic-fetch');
var { API_KEY } = require('./secrets.js');

// TODO: on publish, replace all /dev/ with /v0/ in API calls

// specify your variable values here  
var docType = "auto_insurance_quote"
var docUrl = "https://github.com/sensible-hq/sensible-docs/raw/main/readme-sync/assets/v0/pdfs/auto_insurance_anyco.pdf"
// TODO: on publish, specify API key inline here instead + delete test vars
// var API_KEY = "YOUR_API_KEY"

var extractFromDocUrl = async function() {
    console.log(`Initiating asyn request to extract from doc at url ${docUrl}`)
    var myHeaders = new Headers();
    myHeaders.append("Authorization", `Bearer ${API_KEY}`);
    myHeaders.append("Content-Type", "application/json");

    var raw = JSON.stringify({
        "document_url": docUrl
    });
    var requestOptions = {
        method: 'POST',
        headers: myHeaders,
        body: raw,
        redirect: 'follow'
    };
    let response = await fetch(`https://api.sensible.so/dev/extract_from_url/${docType}`, requestOptions);
    if (!response.ok){
      throw Error(response.statusText);
    };
    let responseJson = await response.json();
    let extractionId = responseJson["id"];
    return extractionId
  }

  var retrieveExtraction = async function(id) {
    console.log(`Retrieving extracted data from extraction id ${id}`);
    // wait a few seconds for the extraction to complete before attempting to retrieve it
    await new Promise(r => setTimeout(r, 3000));
    var myHeaders = new Headers();
    myHeaders.append("Authorization", `Bearer ${API_KEY}`);
    var requestOptions = {
        method: 'GET',
        headers: myHeaders,
        redirect: 'follow'
    };
    let response = await fetch(`https://api.sensible.so/dev/documents/${id}`, requestOptions);
    if (!response.ok){
      throw Error(response.statusText);
    };
    let responseJson = await response.json();

    // to avoid polling in prod for the completed extraction, implement a webhook instead 
    while (responseJson["parsed_document"] == null) {
        console.log(responseJson["status"], "\n");
        response = await fetch(`https://api.sensible.so/dev/documents/${id}`, requestOptions);
        if (!response.ok){
          throw Error(response.statusText);
        };
        responseJson = await response.json();
        if (responseJson["status"] == "FAILED") {
          console.log(JSON.stringify(responseJson, null, 2));
          throw "The extraction failed";
        }
        await new Promise(r => setTimeout(r, 3000));
      }
    console.log("EXTRACTED DATA:");
    console.log(JSON.stringify(responseJson, null, 2));
}

async function main()  {
    let extractionId = await extractFromDocUrl();
    retrieveExtraction(extractionId);
}

if (require.main === module) { 
  main();
}